[cond] add a set_ and data mutation expected failure test

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128457

A follow up of the discussion in https://github.com/pytorch/pytorch/pull/126936.

Cond errors out early because of a graph break triggered by DelayGraphBreakVariable, which is created due to `aten.set_` [here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/tensor.py#L366-L376).

We might need to see what happened to this test if we allow graph break in higher order op.

ydwu4 committed:
[cond] add a set_ and data mutation expected failure test

[ghstack-poisoned]

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -2599,6 +2599,24 @@ def wrapper(x):
 
+    def test_cond_trace_set__and_mutation_with_graph_break(self):
+        def f(a, tmp):
+            a_view = a.view(-1)
+            with torch.no_grad():
+                a.set_(tmp)
+                a_view.mul_(2)
+            return a + tmp
```
@ydwu4 what happens if one does the set_ on an intermediate tensor? i.e. 
```
def f(a, tmp):
    a = a.clone()
    tmp = tmp.clone() 
with torch.no_grad():
                a.set_(tmp)
                a_view.mul_(2)
  return a + tmp
```

ydwu4 replied:
Just did some checking. We're getting a  cannot mutate tensors with frozen storage during functionalization pass in cond. This should be expected? Or we want it to fail at dynamo tracing time?

zou3519 replied:
If we torch.compile the cond:
```
@torch.compile
def f():
   return torch.cond(inp.sum() > 0, f, f, (inp, tmp))
```
does it fail at trace time or at runtime? William's point was that this should fail at trace time

zou3519 replied:
are the failures happening at trace-time or runtime?

ydwu4 replied:
backend = "aot_eager" or "inductor" fails at tracing time.

ydwu4 committed:
Update on "[cond] add a set_ and data mutation expected failure test"


A follow up of the discussion in https://github.com/pytorch/pytorch/pull/126936.

Cond errors out early because of a graph break triggered by DelayGraphBreakVariable, which is created due to `aten.set_` [here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/tensor.py#L366-L376).

We might need to see what happened to this test if we allow graph break in higher order op.



[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] add a set_ and data mutation expected failure test"


A follow up of the discussion in https://github.com/pytorch/pytorch/pull/126936.

Cond errors out early because of a graph break triggered by DelayGraphBreakVariable, which is created due to `aten.set_` [here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/tensor.py#L366-L376).

We might need to see what happened to this test if we allow graph break in higher order op.



[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] add a set_ and data mutation expected failure test"


A follow up of the discussion in https://github.com/pytorch/pytorch/pull/126936.

Cond errors out early because of a graph break triggered by DelayGraphBreakVariable, which is created due to `aten.set_` [here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/tensor.py#L366-L376).

We might need to see what happened to this test if we allow graph break in higher order op.



[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
